cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_ctrl_packages.go

```diff
@@ -238,3 +239,107 @@ func (s *Server) GetAvailablePackageDetail(ctx context.Context, request *corev1.
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to intepret page token %q: %v", request.GetPaginationOptions().GetPageToken(), err)
+	}
+
+	// Assume the default cluster if none is specified
+	if cluster == "" {
+		cluster = s.globalPackagingCluster
+	}
+
+	// retrieve the list of installed packages
+	pkgInstalls, err := s.getPkgInstalls(ctx, cluster, namespace)
+	if err != nil {
+		return nil, errorByStatus("get", "PackageInstall", "", err)
+	}
+
+	// paginate the list of results
+	installedPkgSummaries := make([]*corev1.InstalledPackageSummary, len(pkgInstalls))
+
+	// create the waiting group for processing each item aynchronously
+	var wg sync.WaitGroup
+	if len(pkgInstalls) > 0 {
+		startAt := -1
+		if pageSize > 0 {
+			startAt = int(pageSize) * pageOffset
+		}
+		for i, pkgInstall := range pkgInstalls {
+			wg.Add(1)
+			if startAt <= i {
+				go func(i int, pkgInstall *packagingv1alpha1.PackageInstall) error {
+					defer wg.Done()
+					// fetch additional information from the package metadata
+					// TODO(agamez): if the repository where the package belongs to is not installed, it will throw an error;
+					// decide whether we should ignore it or it is ok with returning an error
+					pkgMetadata, err := s.getPkgMetadata(ctx, cluster, namespace, pkgInstall.Spec.PackageRef.RefName)
+					if err != nil {
+						return errorByStatus("get", "PackageMetadata", pkgInstall.Spec.PackageRef.RefName, err)
+					}
+
+					// Use the field selector to return only Package CRs that match on the spec.refName.
+					fieldSelector := fmt.Sprintf("spec.refName=%s", pkgInstall.Spec.PackageRef.RefName)
+					pkgs, err := s.getPkgsWithFieldSelector(ctx, cluster, namespace, fieldSelector)
+					if err != nil {
+						return errorByStatus("get", "Package", "", err)
+					}
+					pkgVersionsMap, err := getPkgVersionsMap(pkgs)
+					if err != nil {
+						return err
+					}
+
+					// generate the installedPackageSummary from the fetched information
+					installedPackageSummary, err := s.getInstalledPackageSummary(pkgInstall, pkgMetadata, pkgVersionsMap, cluster)
+					if err != nil {
+						return status.Errorf(codes.Internal, fmt.Sprintf("unable to create the InstalledPackageSummary: %v", err))
+					}
+
+					// append the availablePackageSummary to the slice
+					installedPkgSummaries[i] = installedPackageSummary
+
+					return nil
+				}(i, pkgInstall)
+			}
+			// if we've reached the end of the page, stop iterating
+			if pageSize > 0 && len(installedPkgSummaries) == int(pageSize) {
+				break
+			}
+		}
+	}
+	// Wait until each goroutine has finished
+	wg.Wait()
+
+	// TODO(agamez): the slice with make is filled with <nil>, in case of an error in the
+	// i goroutine, the i-th <nil> stub will remain. Check if 'errgroup' works here, but I haven't
+	// been able so far.
+	// An alternative is using channels to perform a fine-grained control... but not sure if it worths
```

Comment:
Yeah. You'd normally avoid multiple go-routines writing to the same slice and instead collect the processed data via a channel as you've noted. But I know you're doing it carefully here so that each go routine is responsible for one index only. Stepping back, I'm not sure why the nil's here would be an issue for you anyway given that we should return an error (and a nil value) if an error has been returned, I'd think?

Reply:
Yes, I'd have liked to dig into the goroutines+channels (it's one of my pending topics in Go; btw, now I have a lot of interesting Greg's code to learn from, so thank you!) but I didn't want to spend more time here (I mean, before landing something we can start iterating from).

> given that we should return an error (and a nil value) if an error has been returned, I'd think?

That's the point, I'm not so sure we should return a "nil, error" if a goroutine (=single package) fails.
See https://github.com/kubeapps/kubeapps/pull/3784#discussion_r756281729

I'm ok reverting back to the hard-error way, but also happy to discuss longer-term ideas like letting users decide or at least returning "values,error", so that the client/UI was able to decide.